src/cargo/core/compiler/context/mod.rs

```diff
@@ -63,6 +65,29 @@ pub struct Unit<'a> {
     pub mode: CompileMode,
 }
 
+impl<'a> Unit<'a> {
+    pub fn buildkey(&self) -> String {
+        #![allow(deprecated)] // for SipHash
+        use std::hash::{Hash, Hasher, SipHasher};
+
+        let mut hasher = SipHasher::new_with_keys(0, 0);
+        self.hash(&mut hasher);
+        format!("{}-{}", self.pkg.name(), hasher.finish())
```

Comment:
I think there's a `short_hash` helper in Cargo which should assist with generating this string